src/groupeddataframe/grouping.jl

```diff
@@ -122,50 +119,111 @@ _names(gd::GroupedDataFrame) = _names(gd.parent)
-  and one row per group.
+- A named tuple of single values or a `DataFrameRow` gives a data frame with one column
+  for each field and one row per group.
 - A vector gives a data frame with a single column and as many rows
   for each group as the length of the returned vector for that group.
-- A data frame or a matrix gives a data frame with the same columns and as many rows
-  for each group as the rows returned for that group.
+- A data frame, a named tuple of vectors or a matrix gives a data frame
+  with the same columns and as many rows for each group as the rows returned for that group.
 
 In all cases, the resulting `GroupedDataFrame` contains all the grouping columns in addition
-to those listed above. Note that `f` must always return the same type of object for
+to those listed above. Column names are automatically generated when necessary: for functions
+operating on a single column and returning a single value or vector, the function name is
+appended to the input colummn name; for other functions, columns are called `x1`, `x2`
+and so on.
+
+Note that `f` must always return the same type of object for
 all groups, and (if a named tuple or data frame) with the same fields or columns.
-Returning a single value or a named tuple is significantly faster than
-returning a vector or a data frame.
+Due to type instability, returning a single value or a named tuple is dramatically
+faster than returning a data frame.
 
 ### Examples
 
-```julia
-df = DataFrame(a = repeat([1, 2, 3, 4], outer=[2]),
-               b = repeat([2, 1], outer=[4]),
-               c = randn(8))
-gd = groupby(df, :a)
-map(d -> sum(skipmissing(d[:c])), gd)
+```jldoctest
+julia> df = DataFrame(a = repeat([1, 2, 3, 4], outer=[2]),
+                      b = repeat([2, 1], outer=[4]),
+                      c = 1:8);
+
+julia> gd = groupby(df, :a);
+
+julia> map(:c => sum, gd)
+GroupedDataFrame{DataFrame} with 4 groups based on key: :a
+First Group: 1 row
+│ Row │ a     │ c_sum │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 6     │
+⋮
+Last Group: 1 row
+│ Row │ a     │ c_sum │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 4     │ 12    │
+
+julia> map(df -> sum(df.c), gd) # Slower variant
+GroupedDataFrame{DataFrame} with 4 groups based on key: :a
+First Group: 1 row
+│ Row │ a     │ x1    │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 6     │
+⋮
+Last Group: 1 row
+│ Row │ a     │ x1    │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 4     │ 12    │
 ```
 
+See [`by`](@ref) for more examples.
+
 ### See also
 
 `combine(f, gd)` returns a `DataFrame` rather than a `GroupedDataFrame`
 
 """
-function Base.map(f::Union{Function, Type}, gd::GroupedDataFrame)
+function Base.map(f::Any, gd::GroupedDataFrame)
     if length(gd) > 0
-        idx, valscat = _combine(wrap(f(gd[1])), f, gd)
+        idx, valscat = _combine(f, gd)
```

Comment:
here `makeunique=false` but in `combine` it is `true` which is inconsistent.

Reply:
Right. I've set `makeunique=true` here too, but let's discuss that at #1460.